ci: add Shellcheck workflow

obnoxxx opened the pull request:
This adds shellcheck linting to the CI.

The purpose of this is to make sure the shell scripts in this repo don't have
syntactic or style issues known to shellcheck and to prevent future PRs from introducing regressions.


It additionally fixes issues in existing scripts while at it to prevent this from breaking the CI right away.

obnoxxx commented:
the new shellcheck workflow is failing. I need to fix more scripts.

rofl0r commented on autogen.sh:
```diff
@@ -1,24 +1,24 @@
 set -x
 
-cd $srcdir
+cd "$srcdir" || exit
```
i'd argue that just exiting without any error message or exit code is worse than running into the aclocal error that would happen when this fails.

obnoxxx replied:
@rofl0r  wrote:
> i'd argue that just exiting without any error message or exit code is worse than running into the aclocal error that would happen when this fails.

Agreed. I changed to have a message along with exit.

rofl0r commented:
is there any specific reason you're pursuing this?
i think i never used autogen.sh, instead always using "autoreconf -i".
as for adding double quotes everywhere my personal opinion is that nobody should use spaces in filenames/paths, and if he/she does, the punishment is just... ;)

obnoxxx commented:
@rofl0r  wrote:
> is there any specific reason you're pursuing this?

I simply wanted to add shellcheck to the CI.

Without breaking ci right away, it requires adjusting some scripts.
That's all.



i think i never used autogen.sh, instead always using "autoreconf -i".

`autogen.sh is at least used in ci workflows. I have been using it as well.

 as for adding double quotes everywhere my personal opinion is that nobody should use spaces in filenames/paths, and if he/she does, the punishment is just... ;)

rofl0r commented on tests/scripts/run_tests.sh:
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/env bash
```
why switch to bash ?

obnoxxx replied:
no particular reason.

iirc, it \seemed to fix something.

Now changing it back to sh.

rofl0r commented on tests/scripts/run_tests.sh:
```diff
@@ -103,85 +105,102 @@ AddHeader "X-My-Header3" "Powered by Tinyproxy"
-	kill $pid
-	if test "x$?" = "x0" ; then
+	printf "killing tinyproxy..."
+    local pid 
```
fwiw, even dash supports local

obnoxxx replied:
yeah, I wasn't removing local. just changing the way it is used. shellcheck said to separate declaration from assignment.

rofl0r commented on autogen.sh:
```diff
@@ -27,4 +33,4 @@ if test $RC -ne 0; then
 
 echo
-echo "Now type 'make' to compile Tinyproxy."
+echo "Nowtype 'make' to compile Tinyproxy."
```
?

obnoxxx replied:
fixed

obnoxxx commented:
@rofl0r:

I have changed the PR to introduce a `shellcheck` make target that can easily be used locally and is now used in the github workflow instead of the action.

I have also split the shellcheck fixes into multiple commits, making sure at each step that the scripts still work as expected (in particular `make test`).

rofl0r commented:
congrats on getting it working.
does the github workflow still enforce the shellcheck to succeed ?
also i'd appreciate if this could be squashed before merging to not unnecessarily clutter the commit history.

obnoxxx commented:
@rofl0r  wrote:
> congrats on getting it working.

👍🏼 

 does the github workflow still enforce the shellcheck to succeed ?

yes, CI would fail if future PRs introduced shellcheck issue.

The main point with the new PR verion is thatthe new  `make shellcheck`  makes it easy and convenient for developers to check it locally before pushing.


>  also i'd appreciate if this could be squashed before merging to not unnecessarily clutter the commit history.

While I am generally in favor of atomic commits, I do agree in this case that there is no value in having multiple commits per file for fixing different shellcheck issues.

I plan to push  an update later today with the   following changes:

* one shellcheck fix commit per file containing fixes for different issues.
* separate commit(s) per file containing syntax fixes and other changes.

rofl0r reviewed (approved):
lgtm
